File: assets/scripts/common/FSM/FSM.ts
export namespace Maze_FSM
{
    export type FSMCallback<StateIdsEnum, Context> = (context:Context) => void;
    type TransitionMap<StateIdsEnum, TransitionIdsEnum, Context> = Map<TransitionIdsEnum, Transition<StateIdsEnum, TransitionIdsEnum, Context>>;

    export class State<StateIdsEnum, TransitionIdsEnum, Context>
    {
        private _stateId:StateIdsEnum;
        public get stateId() : StateIdsEnum
        {
            return this._stateId;
        }

        private _transitionMap:TransitionMap<StateIdsEnum, TransitionIdsEnum, Context>;
        public get transitionMap() : TransitionMap<StateIdsEnum, TransitionIdsEnum, Context>
        {
            return this._transitionMap;
        }
        public set transitionMap(val:TransitionMap<StateIdsEnum, TransitionIdsEnum, Context>)
        {
            this._transitionMap = val;
        }

        addTransition(val:Transition<StateIdsEnum, TransitionIdsEnum, Context>)
        {
            this._transitionMap.set(val.transitionId, val);
        }

        private _onEnter : FSMCallback<StateIdsEnum, Context>;
        private _onExit : FSMCallback<StateIdsEnum, Context>;

        constructor(stateId:StateIdsEnum, onEnter:FSMCallback<StateIdsEnum, Context>, onExit:FSMCallback<StateIdsEnum, Context>)
        {
            this._stateId = stateId;
            this._onEnter = onEnter;
            this._onExit = onExit;
            this._transitionMap = new Map<TransitionIdsEnum, Transition<StateIdsEnum, TransitionIdsEnum, Context>>();
        }

        public enter(context:Context)
        {
            this._onEnter(context);
        }

        public exit(context:Context)
        {
            this._onExit(context);
        }
    }

    export class Transition<StateIdsEnum, TransitionIdsEnum, Context>
    {
        private _onTransition:FSMCallback<StateIdsEnum, Context>;
        
        private _transitionId:TransitionIdsEnum;
        public get transitionId() : TransitionIdsEnum
        {
            return this._transitionId;
        }

        private _startState:State<StateIdsEnum, TransitionIdsEnum, Context>;
        private _endState:State<StateIdsEnum, TransitionIdsEnum, Context>;

        constructor(transitionId:TransitionIdsEnum, 
                    startState:State<StateIdsEnum, TransitionIdsEnum, Context>, 
                    endState:State<StateIdsEnum, TransitionIdsEnum, Context>,
                    onTransition:FSMCallback<StateIdsEnum, Context> = (context:Context)=>{})
        {
            this._transitionId = transitionId;
            this._startState = startState;
            this._endState = endState;
            this._onTransition = onTransition;
        }

        public applyTransition(context:Context) : State<StateIdsEnum, TransitionIdsEnum, Context>
        {
            if(this._startState != this._endState)
            {
                this._startState.exit(context);
                this._endState.enter(context);
            }

            this._onTransition(context);

            return this._endState;
        }
    }

    export class FSM<StateIdsEnum, TransitionIdsEnum, Context>
    {
        private _initialized = false;
        public get initialized()
        {
            return this._initialized;
        }

        private _initialState:State<StateIdsEnum, TransitionIdsEnum, Context>;

        private _currentState:State<StateIdsEnum, TransitionIdsEnum, Context>;
        public get currentState() : StateIdsEnum
        {
            return this._currentState.stateId;
        }
        
        private _context:Context;
        public get context() : Context
        {
            return this._context;
        }
        public set context(val:Context)
        {
            this._context = val;
        }

        constructor( initialState:State<StateIdsEnum, TransitionIdsEnum, Context>, context:Context )
        {
            this._initialState = initialState;
            this._currentState = this._initialState;
            this._context = context;
        }

        init()
        {
            if(false == this._initialized)
            {
                this._initialized = true;
                this._currentState.enter(this._context);
            }
        }

        public applyTransition(transitionId:TransitionIdsEnum)
        {
            if(true == this._initialized)
            {
                var foundTransition = this._currentState.transitionMap.get(transitionId);

                if(null != foundTransition)
                {
                    this._currentState = foundTransition.applyTransition(this._context);
                }
            }
            else
            {
                throw Error("FSM is not initialized!");
            }
        }
    }
}

File: Maze_Unity/Assets/Scripts/Common/FSM/FSM.cs
using System.Collections.Generic;

namespace Maze_FSM
{
    public delegate void FSMCallback<StateIdsEnum, Context>(Context context);

    public class TransitionMap<StateIdsEnum, TransitionIdsEnum, Context> : Dictionary<TransitionIdsEnum, Transition<StateIdsEnum, TransitionIdsEnum, Context>> { }

    public class State<StateIdsEnum, TransitionIdsEnum, Context>
    {
        private StateIdsEnum mStateId;
        public StateIdsEnum StateId { get => mStateId; set => mStateId = value; }

        private TransitionMap<StateIdsEnum, TransitionIdsEnum, Context> mTransitionMap;
        public TransitionMap<StateIdsEnum, TransitionIdsEnum, Context> TransitionMap { get => mTransitionMap; set => mTransitionMap = value; }

        public void AddTransition(Transition<StateIdsEnum, TransitionIdsEnum, Context> val)
        {
            mTransitionMap.Add(val.TransitionId, val);
        }

        private FSMCallback<StateIdsEnum, Context> mOnEnter;
        private FSMCallback<StateIdsEnum, Context> mOnExit;

        public State(StateIdsEnum stateId, FSMCallback<StateIdsEnum, Context> onEnter, FSMCallback<StateIdsEnum, Context> onExit)
        {
            mStateId = stateId;
            mOnEnter = onEnter;
            mOnExit = onExit;
            mTransitionMap = new TransitionMap<StateIdsEnum, TransitionIdsEnum, Context>();
        }

        public void enter(Context context)
        {
            mOnEnter(context);
        }

        public void exit(Context context)
        {
            mOnExit(context);
        }
    }
    public class Transition<StateIdsEnum, TransitionIdsEnum, Context>
    {
        private FSMCallback<StateIdsEnum, Context> mOnTransition;

        private TransitionIdsEnum mTransitionId;
        public TransitionIdsEnum TransitionId { get => mTransitionId; }

        private State<StateIdsEnum, TransitionIdsEnum, Context>  mStartState;
        private State<StateIdsEnum, TransitionIdsEnum, Context>  mEndState;

        public Transition(TransitionIdsEnum transitionId,
                State<StateIdsEnum, TransitionIdsEnum, Context>  startState,
                State<StateIdsEnum, TransitionIdsEnum, Context>  endState,
                FSMCallback<StateIdsEnum, Context>  onTransition = null)
        {
            mTransitionId = transitionId;
            mStartState = startState;
            mEndState = endState;
            mOnTransition = onTransition;
        }

        public State<StateIdsEnum, TransitionIdsEnum, Context>  applyTransition(Context context)
        {
            if (mStartState != mEndState)
            {
                mStartState.exit(context);
                mEndState.enter(context);
            }

            if (null != mOnTransition)
            {
                mOnTransition(context);
            }

            return mEndState;
        }
    }

    public class FSM<StateIdsEnum, TransitionIdsEnum, tContext > where tContext : new()
    {
        private bool mInitialized = false;
        public bool Initialized { get => mInitialized; }

        private State<StateIdsEnum, TransitionIdsEnum, tContext>  mInitialState;
        
        private State<StateIdsEnum, TransitionIdsEnum, tContext>  mCurrentState;
        internal State<StateIdsEnum, TransitionIdsEnum, tContext> CurrentState { get => mCurrentState; set => mCurrentState = value; }

        private tContext mContext;
        public tContext Context { get => mContext; set => mContext = value; }

        public FSM(State<StateIdsEnum, TransitionIdsEnum, tContext> initialState)
        {
            mInitialState = initialState;
            mCurrentState = mInitialState;
            mContext = new tContext();
        }

        // takes effect only if init was already called
        public void restart()
        {
            if (true == mInitialized)
            {
                mContext = new tContext();
                mCurrentState = mInitialState;
                mCurrentState.enter(mContext);
            }
        }

        public void init()
        {
            if (false == mInitialized)
            {
                mInitialized = true;
                mCurrentState.enter(mContext);
            }
        }

        public void ApplyTransition(TransitionIdsEnum transitionId)
        {
            if (true == mInitialized)
            {
                Transition<StateIdsEnum, TransitionIdsEnum, tContext> foundTransition;
                if (mCurrentState.TransitionMap.TryGetValue(transitionId, out foundTransition))
                {
                    mCurrentState = foundTransition.applyTransition(mContext);
                }
            }
            else
            {
                throw new System.Exception("FSM is not initialized!");
            }
        }
    }
}